File: DevDocs/backend/services/document-processing/isin-extractor.js
/**
 * ISIN Extractor
 * 
 * Extracts International Securities Identification Numbers (ISINs) from financial documents
 * and associates them with security names, quantities, and values.
 */

const logger = require('../../utils/logger');

/**
 * Extract ISINs from a document
 * @param {Object} options - ISIN extraction options
 * @param {string} options.extractedText - Extracted text from the document
 * @param {Array} options.tables - Extracted tables from the document
 * @param {Array} options.pages - Array of page objects with text content
 * @returns {Promise<Object>} - ISIN extraction results
 */
async function extractISINs(options) {
  const { extractedText, tables, pages } = options;
  
  // Create result object
  const result = {
    isins: [],
    count: 0
  };
  
  try {
    // Extract ISINs from text
    const textIsins = extractIsinsFromText(extractedText);
    
    // Extract ISINs from tables
    const tableIsins = extractIsinsFromTables(tables);
    
    // Merge and deduplicate ISINs
    const mergedIsins = mergeIsins([...textIsins, ...tableIsins]);
    
    // Enrich ISINs with additional information
    const enrichedIsins = await enrichIsins(mergedIsins, extractedText, tables);
    
    result.isins = enrichedIsins;
    result.count = enrichedIsins.length;
    
    logger.info(`Extracted ${result.count} ISINs from document`);
    return result;
  } catch (error) {
    logger.error(`Error extracting ISINs: ${error.message}`, error);
    throw error;
  }
}

/**
 * Extract ISINs from text using regex
 * @param {string} text - Text to extract ISINs from
 * @returns {Array} - Extracted ISINs
 */
function extractIsinsFromText(text) {
  logger.info(`Extracting ISINs from text`);
  
  const isins = [];
  
  // ISIN pattern: 2 letters followed by 10 alphanumeric characters
  // We look for patterns that are likely to be ISINs in financial documents
  const isinPattern = /\b([A-Z]{2}[A-Z0-9]{10})\b/g;
  
  let match;
  while ((match = isinPattern.exec(text)) !== null) {
    const isin = match[1];
    
    // Validate ISIN
    if (isValidIsin(isin)) {
      // Try to find the security name near the ISIN
      const contextBefore = text.substring(Math.max(0, match.index - 100), match.index);
      const contextAfter = text.substring(match.index + isin.length, Math.min(text.length, match.index + isin.length + 100));
      const context = contextBefore + contextAfter;
      
      const securityName = extractSecurityName(context, isin);
      
      isins.push({
        code: isin,
        name: securityName,
        source: 'text',
        context: context.trim()
      });
    }
  }
  
  logger.info(`Found ${isins.length} ISINs in text`);
  return isins;
}

/**
 * Extract ISINs from tables
 * @param {Array} tables - Tables to extract ISINs from
 * @returns {Array} - Extracted ISINs
 */
function extractIsinsFromTables(tables) {
  logger.info(`Extracting ISINs from ${tables.length} tables`);
  
  const isins = [];
  
  for (const table of tables) {
    // Find the ISIN column index
    const isinColumnIndex = findColumnIndex(table.headers, ['isin', 'security id', 'identifier', 'id']);
    
    if (isinColumnIndex === -1) {
      // No ISIN column found, try to find ISINs in any column
      for (const row of table.rows) {
        for (let i = 0; i < row.length; i++) {
          const cell = row[i];
          const isinMatch = cell.match(/\b([A-Z]{2}[A-Z0-9]{10})\b/);
          
          if (isinMatch && isValidIsin(isinMatch[1])) {
            // Try to find the security name in this row
            const nameColumnIndex = findColumnIndex(table.headers, ['name', 'security', 'description', 'asset']);
            const securityName = nameColumnIndex !== -1 ? row[nameColumnIndex] : '';
            
            // Try to find quantity and value
            const quantityColumnIndex = findColumnIndex(table.headers, ['quantity', 'amount', 'units', 'shares']);
            const valueColumnIndex = findColumnIndex(table.headers, ['value', 'market value', 'total', 'position']);
            
            const quantity = quantityColumnIndex !== -1 ? parseNumber(row[quantityColumnIndex]) : null;
            const value = valueColumnIndex !== -1 ? parseNumber(row[valueColumnIndex]) : null;
            
            isins.push({
              code: isinMatch[1],
              name: securityName,
              quantity: quantity,
              value: value,
              source: 'table',
              tableId: table.id,
              rowIndex: table.rows.indexOf(row)
            });
          }
        }
      }
    } else {
      // ISIN column found, extract ISINs from that column
      for (const row of table.rows) {
        if (isinColumnIndex < row.length) {
          const cell = row[isinColumnIndex];
          const isinMatch = cell.match(/\b([A-Z]{2}[A-Z0-9]{10})\b/);
          
          if (isinMatch && isValidIsin(isinMatch[1])) {
            // Try to find the security name in this row
            const nameColumnIndex = findColumnIndex(table.headers, ['name', 'security', 'description', 'asset']);
            const securityName = nameColumnIndex !== -1 && nameColumnIndex < row.length ? row[nameColumnIndex] : '';
            
            // Try to find quantity and value
            const quantityColumnIndex = findColumnIndex(table.headers, ['quantity', 'amount', 'units', 'shares']);
            const valueColumnIndex = findColumnIndex(table.headers, ['value', 'market value', 'total', 'position']);
            
            const quantity = quantityColumnIndex !== -1 && quantityColumnIndex < row.length ? parseNumber(row[quantityColumnIndex]) : null;
            const value = valueColumnIndex !== -1 && valueColumnIndex < row.length ? parseNumber(row[valueColumnIndex]) : null;
            
            isins.push({
              code: isinMatch[1],
              name: securityName,
              quantity: quantity,
              value: value,
              source: 'table',
              tableId: table.id,
              rowIndex: table.rows.indexOf(row)
            });
          }
        }
      }
    }
  }
  
  logger.info(`Found ${isins.length} ISINs in tables`);
  return isins;
}

/**
 * Find the index of a column in a table based on possible header names
 * @param {Array} headers - Table headers
 * @param {Array} possibleNames - Possible names for the column
 * @returns {number} - Column index or -1 if not found
 */
function findColumnIndex(headers, possibleNames) {
  for (let i = 0; i < headers.length; i++) {
    const header = headers[i].toLowerCase();
    
    for (const name of possibleNames) {
      if (header.includes(name.toLowerCase())) {
        return i;
      }
    }
  }
  
  return -1;
}

/**
 * Parse a number from a string
 * @param {string} str - String to parse
 * @returns {number|null} - Parsed number or null if invalid
 */
function parseNumber(str) {
  if (!str) return null;
  
  // Remove currency symbols, commas, and other non-numeric characters
  const cleanStr = str.replace(/[^\d.-]/g, '');
  
  // Parse as float
  const num = parseFloat(cleanStr);
  
  return isNaN(num) ? null : num;
}

/**
 * Extract security name from context around an ISIN
 * @param {string} context - Text context around the ISIN
 * @param {string} isin - ISIN code
 * @returns {string} - Extracted security name or empty string
 */
function extractSecurityName(context, isin) {
  // Try to find the security name before the ISIN
  // Look for patterns like "Security Name (ISIN)" or "Security Name ISIN"
  const beforePattern = /([A-Za-z0-9\s.,&'-]+)[\s(]*\b${isin}\b/;
  const beforeMatch = context.match(beforePattern);
  
  if (beforeMatch && beforeMatch[1].trim().length > 0) {
    return beforeMatch[1].trim();
  }
  
  // Try to find the security name after the ISIN
  // Look for patterns like "ISIN Security Name" or "ISIN: Security Name"
  const afterPattern = /\b${isin}\b[:\s)]*([A-Za-z0-9\s.,&'-]+)/;
  const afterMatch = context.match(afterPattern);
  
  if (afterMatch && afterMatch[1].trim().length > 0) {
    return afterMatch[1].trim();
  }
  
  return '';
}

/**
 * Merge and deduplicate ISINs
 * @param {Array} isins - Array of ISINs to merge
 * @returns {Array} - Merged and deduplicated ISINs
 */
function mergeIsins(isins) {
  const isinMap = new Map();
  
  for (const isin of isins) {
    const code = isin.code;
    
    if (!isinMap.has(code)) {
      isinMap.set(code, isin);
    } else {
      const existing = isinMap.get(code);
      
      // Merge information, preferring table source over text source
      if (isin.source === 'table' && existing.source === 'text') {
        isinMap.set(code, {
          ...existing,
          ...isin,
          name: isin.name || existing.name
        });
      } else if (isin.source === 'text' && existing.source === 'table') {
        isinMap.set(code, {
          ...existing,
          name: existing.name || isin.name
        });
      } else {
        // Both from same source, merge and take non-null values
        isinMap.set(code, {
          ...existing,
          name: existing.name || isin.name,
          quantity: existing.quantity || isin.quantity,
          value: existing.value || isin.value
        });
      }
    }
  }
  
  return Array.from(isinMap.values());
}

/**
 * Enrich ISINs with additional information
 * @param {Array} isins - ISINs to enrich
 * @param {string} text - Document text
 * @param {Array} tables - Document tables
 * @returns {Promise<Array>} - Enriched ISINs
 */
async function enrichIsins(isins, text, tables) {
  logger.info(`Enriching ${isins.length} ISINs with additional information`);
  
  // For each ISIN, try to find additional information
  for (const isin of isins) {
    // If we don't have a name yet, try harder to find one
    if (!isin.name || isin.name.length === 0) {
      isin.name = findSecurityNameInDocument(isin.code, text, tables);
    }
    
    // If we don't have quantity or value, try to find them
    if (!isin.quantity || !isin.value) {
      const { quantity, value } = findQuantityAndValue(isin.code, text, tables);
      
      if (!isin.quantity) isin.quantity = quantity;
      if (!isin.value) isin.value = value;
    }
    
    // Try to determine the security type
    isin.securityType = determineSecurityType(isin.code, isin.name);
  }
  
  return isins;
}

/**
 * Find security name in the document
 * @param {string} isin - ISIN code
 * @param {string} text - Document text
 * @param {Array} tables - Document tables
 * @returns {string} - Security name or empty string
 */
function findSecurityNameInDocument(isin, text, tables) {
  // Look for the ISIN in the text with more context
  const isinIndex = text.indexOf(isin);
  
  if (isinIndex !== -1) {
    const contextBefore = text.substring(Math.max(0, isinIndex - 200), isinIndex);
    const contextAfter = text.substring(isinIndex + isin.length, Math.min(text.length, isinIndex + isin.length + 200));
    const context = contextBefore + contextAfter;
    
    // Look for patterns that might indicate a security name
    const patterns = [
      // Name before ISIN
      new RegExp(`([A-Za-z0-9\\s.,&'-]{5,50})\\s*\\(\\s*${isin}\\s*\\)`, 'i'),
      new RegExp(`([A-Za-z0-9\\s.,&'-]{5,50})\\s*${isin}`, 'i'),
      // Name after ISIN
      new RegExp(`${isin}\\s*:\\s*([A-Za-z0-9\\s.,&'-]{5,50})`, 'i'),
      new RegExp(`${isin}\\s*([A-Za-z0-9\\s.,&'-]{5,50})`, 'i')
    ];
    
    for (const pattern of patterns) {
      const match = context.match(pattern);
      if (match && match[1].trim().length > 0) {
        return match[1].trim();
      }
    }
  }
  
  // Look in tables for rows that contain this ISIN
  for (const table of tables) {
    for (const row of table.rows) {
      const rowText = row.join(' ');
      
      if (rowText.includes(isin)) {
        // Try to find a cell that might contain a name
        // Usually it's one of the first few cells
        for (let i = 0; i < Math.min(3, row.length); i++) {
          const cell = row[i];
          
          // Check if this cell doesn't contain the ISIN and has a reasonable length
          if (!cell.includes(isin) && cell.length > 3 && cell.length < 100) {
            return cell.trim();
          }
        }
      }
    }
  }
  
  return '';
}

/**
 * Find quantity and value for an ISIN
 * @param {string} isin - ISIN code
 * @param {string} text - Document text
 * @param {Array} tables - Document tables
 * @returns {Object} - Object with quantity and value
 */
function findQuantityAndValue(isin, text, tables) {
  let quantity = null;
  let value = null;
  
  // Look in tables for rows that contain this ISIN
  for (const table of tables) {
    for (const row of table.rows) {
      const rowText = row.join(' ');
      
      if (rowText.includes(isin)) {
        // Try to find quantity and value columns
        const quantityColumnIndex = findColumnIndex(table.headers, ['quantity', 'amount', 'units', 'shares']);
        const valueColumnIndex = findColumnIndex(table.headers, ['value', 'market value', 'total', 'position']);
        
        if (quantityColumnIndex !== -1 && quantityColumnIndex < row.length) {
          quantity = parseNumber(row[quantityColumnIndex]);
        }
        
        if (valueColumnIndex !== -1 && valueColumnIndex < row.length) {
          value = parseNumber(row[valueColumnIndex]);
        }
        
        if (quantity !== null && value !== null) {
          return { quantity, value };
        }
      }
    }
  }
  
  // If we couldn't find both in tables, try to find them in text
  if (quantity === null || value === null) {
    const isinIndex = text.indexOf(isin);
    
    if (isinIndex !== -1) {
      const contextAfter = text.substring(isinIndex + isin.length, Math.min(text.length, isinIndex + isin.length + 300));
      
      // Look for quantity pattern
      if (quantity === null) {
        const quantityPattern = /(?:quantity|amount|units|shares)[:\s]*([0-9,.]+)/i;
        const quantityMatch = contextAfter.match(quantityPattern);
        
        if (quantityMatch) {
          quantity = parseNumber(quantityMatch[1]);
        }
      }
      
      // Look for value pattern
      if (value === null) {
        const valuePattern = /(?:value|market value|total|position)[:\s]*([0-9,.]+)/i;
        const valueMatch = contextAfter.match(valuePattern);
        
        if (valueMatch) {
          value = parseNumber(valueMatch[1]);
        }
      }
    }
  }
  
  return { quantity, value };
}

/**
 * Determine the security type based on ISIN and name
 * @param {string} isin - ISIN code
 * @param {string} name - Security name
 * @returns {string} - Security type
 */
function determineSecurityType(isin, name) {
  // ISIN country code (first two letters)
  const countryCode = isin.substring(0, 2);
  
  // Check name for common security type indicators
  const nameLower = name.toLowerCase();
  
  if (nameLower.includes('bond') || nameLower.includes('note') || nameLower.includes('treasury') || 
      nameLower.includes('gilt') || nameLower.includes('debenture')) {
    return 'bond';
  } else if (nameLower.includes('share') || nameLower.includes('stock') || 
             nameLower.includes('common') || nameLower.includes('preferred')) {
    return 'equity';
  } else if (nameLower.includes('fund') || nameLower.includes('etf') || 
             nameLower.includes('trust') || nameLower.includes('index')) {
    return 'fund';
  } else if (nameLower.includes('warrant') || nameLower.includes('right') || 
             nameLower.includes('option')) {
    return 'derivative';
  } else if (nameLower.includes('structured') || nameLower.includes('certificate') || 
             nameLower.includes('note')) {
    return 'structured_product';
  }
  
  // Default to 'unknown'
  return 'unknown';
}

/**
 * Validate an ISIN code
 * @param {string} isin - ISIN code to validate
 * @returns {boolean} - Whether the ISIN is valid
 */
function isValidIsin(isin) {
  // Basic format check: 2 letters followed by 10 alphanumeric characters
  if (!/^[A-Z]{2}[A-Z0-9]{10}$/.test(isin)) {
    return false;
  }
  
  // Check digit validation
  const isinWithoutCheckDigit = isin.slice(0, -1);
  const checkDigit = parseInt(isin.slice(-1), 36);
  
  // Convert letters to numbers (A=10, B=11, ..., Z=35)
  let expandedIsin = '';
  for (let i = 0; i < isinWithoutCheckDigit.length; i++) {
    const char = isinWithoutCheckDigit[i];
    const code = char.charCodeAt(0);
    
    if (code >= 65 && code <= 90) {
      // Letter A-Z
      expandedIsin += (code - 55).toString();
    } else {
      // Digit 0-9
      expandedIsin += char;
    }
  }
  
  // Apply Luhn algorithm
  let sum = 0;
  for (let i = 0; i < expandedIsin.length; i++) {
    let digit = parseInt(expandedIsin[i], 10);
    
    if (i % 2 === 0) {
      digit *= 2;
      if (digit > 9) {
        digit -= 9;
      }
    }
    
    sum += digit;
  }
  
  const calculatedCheckDigit = (10 - (sum % 10)) % 10;
  
  return checkDigit === calculatedCheckDigit;
}

module.exports = {
  extractISINs
};

File: backv2-github/DevDocs/FinDocRAG/extractors/isin_extractor.py
"""
ISIN extractor for financial document processing.
"""
import logging
import re
import pandas as pd
from typing import List, Dict, Any, Optional

logger = logging.getLogger(__name__)

class ISINExtractor:
    """
    Extract ISINs and financial data from documents.
    """
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        """
        Initialize the ISIN extractor.
        
        Args:
            config: Configuration options
        """
        self.config = config or {}
    
    def extract_isins(self, text: str) -> List[str]:
        """
        Extract ISINs from text.
        
        Args:
            text: Text to extract ISINs from
            
        Returns:
            List of extracted ISINs
        """
        # ISIN pattern: 2 letters followed by 10 alphanumeric characters
        isin_pattern = r'[A-Z]{2}[A-Z0-9]{10}'
        
        # Find all matches
        isins = re.findall(isin_pattern, text)
        
        # Remove duplicates
        unique_isins = list(set(isins))
        
        # Validate ISINs
        valid_isins = [isin for isin in unique_isins if self._validate_isin(isin)]
        
        return valid_isins
    
    def extract_financial_data(self, text: str, tables: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Extract financial data from text and tables.
        
        Args:
            text: Text to extract financial data from
            tables: Tables extracted from the document
            
        Returns:
            Extracted financial data
        """
        # Extract ISINs
        isins = self.extract_isins(text)
        
        # Extract securities
        securities = self._extract_securities(text, tables, isins)
        
        # Extract total value
        total_value = self._extract_total_value(text, tables)
        
        # Extract currency
        currency = self._extract_currency(text, tables)
        
        # Extract asset allocation
        asset_allocation = self._extract_asset_allocation(text, tables)
        
        # Compile financial data
        financial_data = {
            "securities": securities,
            "total_value": total_value,
            "currency": currency,
            "asset_allocation": asset_allocation,
            "metrics": {}
        }
        
        return financial_data
    
    def extract_financial_data_from_excel(self, sheets_data: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Extract financial data from Excel sheets.
        
        Args:
            sheets_data: Data from Excel sheets
            
        Returns:
            Extracted financial data
        """
        # Extract text from all sheets
        all_text = " ".join([sheet["text"] for sheet in sheets_data])
        
        # Extract ISINs
        isins = self.extract_isins(all_text)
        
        # Extract securities
        securities = []
        
        for sheet in sheets_data:
            if "data" in sheet:
                for row in sheet["data"]:
                    # Check if any cell contains an ISIN
                    for cell in row.values():
                        for isin in isins:
                            if isin in str(cell):
                                # Extract security data
                                security = self._extract_security_from_row(row, isin)
                                if security:
                                    securities.append(security)
        
        # Extract total value
        total_value = 0
        
        for sheet in sheets_data:
            if "data" in sheet:
                for row in sheet["data"]:
                    # Look for total value
                    for key, value in row.items():
                        if "total" in str(key).lower() and self._is_numeric(value):
                            try:
                                total_value = float(self._clean_numeric(value))
                                break
                            except ValueError:
                                pass
        
        # Extract currency
        currency = self._extract_currency_from_excel(sheets_data)
        
        # Compile financial data
        financial_data = {
            "securities": securities,
            "total_value": total_value,
            "currency": currency,
            "asset_allocation": {},
            "metrics": {}
        }
        
        return financial_data
    
    def extract_financial_data_from_csv(self, csv_data: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Extract financial data from CSV data.
        
        Args:
            csv_data: Data from CSV file
            
        Returns:
            Extracted financial data
        """
        # Convert CSV data to text
        text = ""
        for row in csv_data:
            text += " ".join(str(value) for value in row.values()) + " "
        
        # Extract ISINs
        isins = self.extract_isins(text)
        
        # Extract securities
        securities = []
        
        for row in csv_data:
            # Check if any cell contains an ISIN
            for cell in row.values():
                for isin in isins:
                    if isin in str(cell):
                        # Extract security data
                        security = self._extract_security_from_row(row, isin)
                        if security:
                            securities.append(security)
        
        # Extract total value
        total_value = 0
        
        for row in csv_data:
            # Look for total value
            for key, value in row.items():
                if "total" in str(key).lower() and self._is_numeric(value):
                    try:
                        total_value = float(self._clean_numeric(value))
                        break
                    except ValueError:
                        pass
        
        # Extract currency
        currency = self._extract_currency_from_csv(csv_data)
        
        # Compile financial data
        financial_data = {
            "securities": securities,
            "total_value": total_value,
            "currency": currency,
            "asset_allocation": {},
            "metrics": {}
        }
        
        return financial_data
    
    def _validate_isin(self, isin: str) -> bool:
        """
        Validate an ISIN.
        
        Args:
            isin: ISIN to validate
            
        Returns:
            True if valid, False otherwise
        """
        # Basic validation: 2 letters followed by 10 alphanumeric characters
        if not re.match(r'^[A-Z]{2}[A-Z0-9]{10}$', isin):
            return False
        
        # Check digit validation (Luhn algorithm)
        try:
            # Convert letters to numbers (A=10, B=11, ..., Z=35)
            digits = []
            for char in isin:
                if char.isalpha():
                    digits.append(ord(char) - ord('A') + 10)
                else:
                    digits.append(int(char))
            
            # Reverse the digits
            digits = digits[::-1]
            
            # Apply Luhn algorithm
            total = 0
            for i, digit in enumerate(digits):
                if i % 2 == 0:
                    total += digit
                else:
                    total += sum(int(d) for d in str(digit * 2))
            
            return total % 10 == 0
        except Exception:
            return False
    
    def _extract_securities(self, text: str, tables: List[Dict[str, Any]], isins: List[str]) -> List[Dict[str, Any]]:
        """
        Extract securities from text and tables.
        
        Args:
            text: Text to extract securities from
            tables: Tables extracted from the document
            isins: List of ISINs
            
        Returns:
            List of extracted securities
        """
        securities = []
        
        # Extract securities from tables
        for table in tables:
            if "data" not in table:
                continue
            
            for row in table["data"]:
                # Check if any cell contains an ISIN
                for cell in row.values():
                    for isin in isins:
                        if isin in str(cell):
                            # Extract security data
                            security = self._extract_security_from_row(row, isin)
                            if security:
                                securities.append(security)
        
        # Extract securities from text
        for isin in isins:
            # Find context around ISIN
            isin_index = text.find(isin)
            if isin_index >= 0:
                # Get context (100 characters before and after)
                start = max(0, isin_index - 100)
                end = min(len(text), isin_index + 100)
                context = text[start:end]
                
                # Extract security data
                security = self._extract_security_from_text(context, isin)
                if security:
                    # Check if already extracted
                    if not any(s["identifier"] == isin for s in securities):
                        securities.append(security)
        
        return securities
    
    def _extract_security_from_row(self, row: Dict[str, Any], isin: str) -> Optional[Dict[str, Any]]:
        """
        Extract security data from a table row.
        
        Args:
            row: Table row
            isin: ISIN
            
        Returns:
            Extracted security data or None
        """
        security = {
            "name": "",
            "identifier": isin,
            "quantity": None,
            "value": None
        }
        
        # Extract name
        for key, value in row.items():
            if "name" in str(key).lower() or "description" in str(key).lower():
                security["name"] = str(value).strip()
                break
        
        # If no name found, use the first non-empty cell
        if not security["name"]:
            for value in row.values():
                if value and not self._is_numeric(value) and isin not in str(value):
                    security["name"] = str(value).strip()
                    break
        
        # Extract quantity
        for key, value in row.items():
            if "quantity" in str(key).lower() or "amount" in str(key).lower() or "units" in str(key).lower():
                if self._is_numeric(value):
                    security["quantity"] = self._clean_numeric(value)
                    break
        
        # Extract value
        for key, value in row.items():
            if "value" in str(key).lower() or "total" in str(key).lower() or "market" in str(key).lower():
                if self._is_numeric(value):
                    security["value"] = self._clean_numeric(value)
                    break
        
        # If no value found, look for the last numeric cell
        if security["value"] is None:
            for value in reversed(list(row.values())):
                if self._is_numeric(value):
                    security["value"] = self._clean_numeric(value)
                    break
        
        return security
    
    def _extract_security_from_text(self, text: str, isin: str) -> Optional[Dict[str, Any]]:
        """
        Extract security data from text.
        
        Args:
            text: Text to extract security data from
            isin: ISIN
            
        Returns:
            Extracted security data or None
        """
        security = {
            "name": "",
            "identifier": isin,
            "quantity": None,
            "value": None
        }
        
        # Extract name
        # Look for common patterns
        name_patterns = [
            r'([^0-9\n]{3,50})\s+' + re.escape(isin),  # Name before ISIN
            re.escape(isin) + r'\s+([^0-9\n]{3,50})'   # Name after ISIN
        ]
        
        for pattern in name_patterns:
            match = re.search(pattern, text)
            if match:
                security["name"] = match.group(1).strip()
                break
        
        # Extract quantity and value
        # Look for numbers near the ISIN
        number_pattern = r'[\d,\.\']+\s*(?:[A-Z]{3})?'
        numbers = re.findall(number_pattern, text)
        
        if len(numbers) >= 2:
            # Assume first number is quantity, last is value
            try:
                security["quantity"] = self._clean_numeric(numbers[0])
                security["value"] = self._clean_numeric(numbers[-1])
            except ValueError:
                pass
        
        return security
    
    def _extract_total_value(self, text: str, tables: List[Dict[str, Any]]) -> float:
        """
        Extract total value from text and tables.
        
        Args:
            text: Text to extract total value from
            tables: Tables extracted from the document
            
        Returns:
            Extracted total value
        """
        # Look for total value in tables
        for table in tables:
            if "data" not in table:
                continue
            
            for row in table["data"]:
                # Check if row contains "total"
                for key, value in row.items():
                    if "total" in str(key).lower() or "total" in str(value).lower():
                        # Look for numeric value
                        for cell in row.values():
                            if self._is_numeric(cell):
                                try:
                                    return float(self._clean_numeric(cell))
                                except ValueError:
                                    pass
        
        # Look for total value in text
        total_patterns = [
            r'total\s+value\s*[:\-]?\s*([\d,\.\']+)',
            r'total\s*[:\-]?\s*([\d,\.\']+)',
            r'portfolio\s+value\s*[:\-]?\s*([\d,\.\']+)'
        ]
        
        for pattern in total_patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                try:
                    return float(self._clean_numeric(match.group(1)))
                except ValueError:
                    pass
        
        return 0.0
    
    def _extract_currency(self, text: str, tables: List[Dict[str, Any]]) -> str:
        """
        Extract currency from text and tables.
        
        Args:
            text: Text to extract currency from
            tables: Tables extracted from the document
            
        Returns:
            Extracted currency
        """
        # Common currencies
        currencies = ["USD", "EUR", "GBP", "CHF", "JPY", "CAD", "AUD"]
        
        # Look for currency in tables
        for table in tables:
            if "data" not in table:
                continue
            
            for row in table["data"]:
                for cell in row.values():
                    for currency in currencies:
                        if currency in str(cell):
                            return currency
        
        # Look for currency in text
        for currency in currencies:
            if currency in text:
                return currency
        
        # Look for currency symbols
        currency_symbols = {
            "$": "USD",
            "€": "EUR",
            "£": "GBP",
            "¥": "JPY",
            "Fr.": "CHF"
        }
        
        for symbol, currency in currency_symbols.items():
            if symbol in text:
                return currency
        
        return "USD"  # Default
    
    def _extract_asset_allocation(self, text: str, tables: List[Dict[str, Any]]) -> Dict[str, float]:
        """
        Extract asset allocation from text and tables.
        
        Args:
            text: Text to extract asset allocation from
            tables: Tables extracted from the document
            
        Returns:
            Extracted asset allocation
        """
        asset_allocation = {}
        
        # Common asset classes
        asset_classes = [
            "Equities", "Bonds", "Cash", "Commodities", "Real Estate",
            "Alternatives", "Fixed Income", "Stocks", "Funds", "ETFs",
            "Derivatives", "Options", "Futures", "Currencies", "Crypto",
            "Structured Products", "Liquidity", "Other"
        ]
        
        # Look for asset allocation in tables
        for table in tables:
            if "data" not in table:
                continue
            
            # Check if table might be an asset allocation table
            is_allocation_table = False
            for row in table["data"]:
                for cell in row.values():
                    if "allocation" in str(cell).lower() or "asset" in str(cell).lower():
                        is_allocation_table = True
                        break
                if is_allocation_table:
                    break
            
            if is_allocation_table:
                for row in table["data"]:
                    # Look for asset class and percentage
                    asset_class = None
                    percentage = None
                    
                    for cell in row.values():
                        # Check if cell contains an asset class
                        for ac in asset_classes:
                            if ac.lower() in str(cell).lower():
                                asset_class = ac
                                break
                        
                        # Check if cell contains a percentage
                        if "%" in str(cell):
                            try:
                                percentage = float(str(cell).replace("%", "").strip())
                            except ValueError:
                                pass
                    
                    if asset_class and percentage is not None:
                        asset_allocation[asset_class] = percentage
        
        # If no asset allocation found, look for it in text
        if not asset_allocation:
            for asset_class in asset_classes:
                pattern = r'{}[^%]*?(\d+\.?\d*)%'.format(asset_class)
                match = re.search(pattern, text, re.IGNORECASE)
                if match:
                    try:
                        percentage = float(match.group(1))
                        asset_allocation[asset_class] = percentage
                    except ValueError:
                        pass
        
        return asset_allocation
    
    def _extract_currency_from_excel(self, sheets_data: List[Dict[str, Any]]) -> str:
        """
        Extract currency from Excel sheets.
        
        Args:
            sheets_data: Data from Excel sheets
            
        Returns:
            Extracted currency
        """
        # Common currencies
        currencies = ["USD", "EUR", "GBP", "CHF", "JPY", "CAD", "AUD"]
        
        # Look for currency in sheets
        for sheet in sheets_data:
            if "data" not in sheet:
                continue
            
            for row in sheet["data"]:
                for cell in row.values():
                    for currency in currencies:
                        if currency in str(cell):
                            return currency
        
        return "USD"  # Default
    
    def _extract_currency_from_csv(self, csv_data: List[Dict[str, Any]]) -> str:
        """
        Extract currency from CSV data.
        
        Args:
            csv_data: Data from CSV file
            
        Returns:
            Extracted currency
        """
        # Common currencies
        currencies = ["USD", "EUR", "GBP", "CHF", "JPY", "CAD", "AUD"]
        
        # Look for currency in CSV data
        for row in csv_data:
            for cell in row.values():
                for currency in currencies:
                    if currency in str(cell):
                        return currency
        
        return "USD"  # Default
    
    def _is_numeric(self, value: Any) -> bool:
        """
        Check if a value is numeric.
        
        Args:
            value: Value to check
            
        Returns:
            True if numeric, False otherwise
        """
        if value is None:
            return False
        
        # Convert to string
        value_str = str(value)
        
        # Remove common formatting
        value_str = value_str.replace(",", "").replace("'", "").replace(" ", "")
        
        # Check if it's a number
        try:
            float(value_str)
            return True
        except ValueError:
            return False
    
    def _clean_numeric(self, value: Any) -> float:
        """
        Clean a numeric value.
        
        Args:
            value: Value to clean
            
        Returns:
            Cleaned numeric value
        """
        if value is None:
            return 0.0
        
        # Convert to string
        value_str = str(value)
        
        # Remove common formatting
        value_str = value_str.replace(",", "").replace("'", "").replace(" ", "")
        
        # Remove currency symbols
        currency_symbols = ["$", "€", "£", "¥", "Fr."]
        for symbol in currency_symbols:
            value_str = value_str.replace(symbol, "")
        
        # Remove percentage sign
        value_str = value_str.replace("%", "")
        
        # Convert to float
        return float(value_str)
